web-wallet/src/lib/stores/walletStore.js

```diff
@@ -114,54 +112,89 @@ async function init(wallet) {
     currentAddress,
     initialized: true,
   });
-  sync();
+
+  sync(syncFromBlock).then(() => {
+    settingsStore.update(setKey("userId", currentAddress));
+  });
 }
 
 /** @type {WalletStoreServices["setCurrentAddress"]} */
 async function setCurrentAddress(address) {
   const store = get(walletStore);
 
   return store.addresses.includes(address)
-    ? Promise.resolve(set({ ...store, currentAddress: address })).then(sync)
+    ? Promise.resolve(set({ ...store, currentAddress: address })).then(() =>
+        sync()
+      )
     : Promise.reject(new Error("The received address is not in the list"));
 }
 
 /** @type {WalletStoreServices["stake"]} */
-
-const stake = async (amount, gasPrice, gasLimit) =>
+const stake = async (amount, gasSettings) =>
   syncedAction(() => {
     // @ts-expect-error
-    walletInstance.gasLimit = gasLimit;
-
-    // @ts-expect-error
-    walletInstance.gasPrice = gasPrice;
-
-    // @ts-expect-error
-    return walletInstance.stake(getCurrentAddress(), amount);
+    return walletInstance.stake(
+      getCurrentAddress(),
+      amount,
+      new Gas(gasSettings)
+    );
   });
 
 /** @type {WalletStoreServices["sync"]} */
-function sync() {
+function sync(from) {
   if (!walletInstance) {
     throw new Error("No wallet instance to sync");
   }
 
   if (!syncPromise) {
     const store = get(walletStore);
 
-    set({ ...store, error: null, isSyncing: true });
+    set({
+      ...store,
+      syncStatus: {
+        isInProgress: true,
+        current: 0,
+        last: 0,
+        error: null,
+      },
+    });
 
     syncController = new AbortController();
+
+    const syncOptions = {
+      from,
+      signal: syncController.signal,
+      onblock: (/** @type {number} */ current, /** @type {number} */ last) => {
```

Comment:
Maybe it would be clearer by adding a `@type {(current: number, last: number) => void}` before instead of mixing code and comments?